core/src/main/java/io/temporal/samples/hello/HelloAccumulator.java

```diff
@@ -0,0 +1,625 @@
+    // workflow receives an exit signal.
+    @SignalMethod
+    void exit();
+  }
+
+  /**
+   * This is the Activity Definition's Interface. Activities are building blocks of any Temporal
+   * Workflow and contain any business logic that could perform long running computation, network
+   * calls, etc.
+   *
+   * <p>Annotating Activity Definition methods with @ActivityMethod is optional.
+   *
+   * @see io.temporal.activity.ActivityInterface
+   * @see io.temporal.activity.ActivityMethod
+   */
+  @ActivityInterface
+  public interface GreetingActivities {
+    String composeGreeting(ArrayDeque<Greeting> greetings);
+  }
+
+  /** Simple activity implementation. */
+  static class GreetingActivitiesImpl implements GreetingActivities {
+
+    // here is where we process all of the signals together
+    @Override
+    public String composeGreeting(ArrayDeque<Greeting> greetings) {
+      List<String> greetingList =
+          greetings.stream().map(u -> u.greetingText).collect(Collectors.toList());
+      return "Hello (" + greetingList.size() + ") robots: " + greetingList + "!";
+    }
+  }
+
+  // Main workflow method
+  public static class AccumulatorWorkflowImpl implements AccumulatorWorkflow {
+
+    private final GreetingActivities activities =
+        Workflow.newActivityStub(
+            GreetingActivities.class,
+            ActivityOptions.newBuilder().setStartToCloseTimeout(Duration.ofSeconds(2)).build());
+
+    private static final Logger logger = LoggerFactory.getLogger(AccumulatorWorkflowImpl.class);
+    String bucketKey;
+    ArrayDeque<Greeting> greetings;
+    HashSet<String> allGreetingsSet;
+    boolean stopWaiting = false;
+    ArrayDeque<Greeting> unprocessedGreetings;
+
+    @Override
+    public String accumulateGreetings(
+        String bucketKeyInput,
+        ArrayDeque<Greeting> greetingsInput,
+        Set<String> allGreetingsSetInput) {
+      bucketKey = bucketKeyInput;
+      if (unprocessedGreetings == null) unprocessedGreetings = new ArrayDeque<Greeting>();
+      greetings = new ArrayDeque<Greeting>();
+      allGreetingsSet = new HashSet<String>();
+      greetings.addAll(greetingsInput);
+      allGreetingsSet.addAll(allGreetingsSetInput);
+
+      // If you want to wait for a fixed amount of time instead of a time after a message
+      // as this does now, you might want to check out
+      // ../../updatabletimer
+
+      // Main Workflow Loop:
+      //    - wait for signals to come in
+      //    - every time a signal comes in, wait again for MAX_AWAIT_TIME
+      //    - if time runs out, and there are no messages, process them all and exit
+      //    - if exit signal is received, process any remaining signals and exit
+      do {
+
+        Workflow.await(MAX_AWAIT_TIME, () -> !unprocessedGreetings.isEmpty() || stopWaiting);
+
+        if (!unprocessedGreetings.isEmpty()) {
+          processGreeting(unprocessedGreetings.removeFirst());
```

Comment:
It's a good question. Do you mean, many process more than one greeting instead of just one here per loop? Or am I missing what you mean by "outside this loop"?